services/web/server/src/simcore_service_webserver/garbage_collector/_core_orphans.py

```diff
@@ -1,180 +1,124 @@
+    # otherwise if some time goes it could very well be that there are new projects opened
+    # in between and the GC would remove services that actually should be running.
 
-        project_uuid = resources["project_id"]
-        node_ids: set[str] = await get_workbench_node_ids_from_project_uuid(
-            app, project_uuid
+    with log_catch(_logger, reraise=False):
+        running_services = await director_v2_api.list_dynamic_services(app)
+        if not running_services:
+            # nothing to do
+            return
+        _logger.debug(
+            "Currently running dynamic services: %s",
+            [(x.node_uuid, x.host) for x in running_services],
+        )
+        running_services_by_id: dict[NodeID, DynamicServiceGet] = {
+            service.node_uuid: service for service in running_services
+        }
+
+        known_opened_project_ids = await _list_opened_project_ids(registry)
+        potentially_running_service_ids: list[
+            set[NodeID] | BaseException
+        ] = await logged_gather(
+            *(list_node_ids_in_project(app, _) for _ in known_opened_project_ids),
+            log=_logger,
+            max_concurrency=_MAX_CONCURRENT_CALLS,
+            reraise=False,
+        )
+        potentially_running_service_ids_set: set[NodeID] = set().union(
+            *(_ for _ in potentially_running_service_ids if isinstance(_, set))
         )
-        for node_id in node_ids:
-            currently_opened_projects_node_ids[node_id] = project_uuid
-
-    running_dynamic_services: list[dict[str, Any]] = []
-    try:
-        running_dynamic_services = await director_v2_api.list_dynamic_services(app)
-    except director_v2_api.DirectorServiceError:
-        _logger.debug("Could not fetch list_dynamic_services")
-
-    _logger.info(
-        "Observing running services (this does not mean I am removing them) %s",
-        [
-            (x.get("service_uuid", ""), x.get("service_host", ""))
-            for x in running_dynamic_services
-        ],
-    )
-
-    # if there are multiple dynamic services to stop,
-    # this ensures they are being stopped in parallel
-    # when the user is timed out and the GC needs to close
-    # a big study with logs of heavy projects, this will
-    # ensure it gets done in parallel
-    tasks = [
-        _remove_single_service_if_orphan(
-            app, service, currently_opened_projects_node_ids
+        _logger.debug(
+            "Currently known opened node ids: %s", potentially_running_service_ids_set
         )
-        for service in running_dynamic_services
-    ]
-    await logged_gather(*tasks, reraise=False)
 
-    _logger.debug("Finished orphaned services removal")
+        # compute the difference to find the orphaned services
+        orphaned_running_service_ids = (
+            set(running_services_by_id) - potentially_running_service_ids_set
+        )
```

Comment:
What happens if a service is running but then closed before `potentially_running_service_ids_set` is computed? I.e. a service is marked as orphaned but is actually closed. Will the GC complain when trying to close a service which is already closed?

Reply:
I think not. but I will check. good point